### ISSUE TITLE
issues building with libressl-3.0.1 (wget2-1.99.2)

### ISSUE THREAD

**Issue by mssx86** [rockdaboot/wget2#213]
when configured with:
```
./configure \
 --host=x86_64-linux-musl \
 --build=x86_64-linux-musl \
 --prefix=/usr \
 --sysconfdir=/etc \
 --disable-static \
 --with-ssl=libressl
```
wget2 builds without an error but then prints ```Failed to connect: libwget has been built without TLS support``` when i try to pull anything from an https address.

----------------------------

when configured with:
```
./configure \
 --host=x86_64-linux-musl \
 --build=x86_64-linux-musl \
 --prefix=/usr \
 --sysconfdir=/etc \
 --disable-static \
 --with-openssl
```
it still builds with `gnutls` and running `ldd` on `/usr/bin/wget2` or `/usr/lib/libwget.so` does not return `libssl`. also `-lssl` is not displayed here in the prompt after the `configure` script is run:
```
Libs:               -lpcre2-8 -lidn -llzma -lz -lgpgme -lassuan -lgpg-error -lbz2 -lgnutls
```

----------------------------

when configured with:
```
./configure \
 --host=x86_64-linux-musl \
 --build=x86_64-linux-musl \
 --prefix=/usr \
 --sysconfdir=/etc \
 --disable-static \
 --with-openssl \
 --with-ssl=openssl
```
the build fails with this error:
```
ssl_openssl.c: In function 'openssl_set_priorities':
ssl_openssl.c:340:37: error: 'TLS_MAX_VERSION' undeclared (first use in this function); did you mean 'TLS1_VERSION'?
  340 |  SSL_CTX_set_max_proto_version(ctx, TLS_MAX_VERSION);
      |                                     ^~~~~~~~~~~~~~~
      |                                     TLS1_VERSION
ssl_openssl.c:340:37: note: each undeclared identifier is reported only once for each function it appears in
ssl_openssl.c:355:24: error: 'TLS1_3_VERSION' undeclared (first use in this function); did you mean 'TLS1_2_VERSION'?
  355 |   SET_MIN_VERSION(ctx, TLS1_3_VERSION);
      |                        ^~~~~~~~~~~~~~
ssl_openssl.c:328:42: note: in definition of macro 'SET_MIN_VERSION'
  328 |  if (!SSL_CTX_set_min_proto_version(ctx, ver)) \
      |                                          ^~~
  CCLD     libcsstokenizer.la
ssl_openssl.c: In function 'ssl_resume_session':
ssl_openssl.c:683:8: warning: implicit declaration of function 'SSL_SESSION_is_resumable' [-Wimplicit-function-declaration]
  683 |   if (!SSL_SESSION_is_resumable(ssl_session))
      |        ^~~~~~~~~~~~~~~~~~~~~~~~
ssl_openssl.c: In function 'wget_ssl_open':
ssl_openssl.c:834:3: warning: implicit declaration of function 'SSL_set_hostflags'; did you mean 'SSL_set_state'? [-Wimplicit-function-declaration]
  834 |   SSL_set_hostflags(ssl, X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS);
      |   ^~~~~~~~~~~~~~~~~
      |   SSL_set_state
ssl_openssl.c:836:26: error: 'X509_CHECK_FLAG_NEVER_CHECK_SUBJECT' undeclared (first use in this function); did you mean 'X509_CHECK_FLAG_ALWAYS_CHECK_SUBJECT'?
  836 |   SSL_set_hostflags(ssl, X509_CHECK_FLAG_NEVER_CHECK_SUBJECT);
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                          X509_CHECK_FLAG_ALWAYS_CHECK_SUBJECT
make[2]: *** [Makefile:2214: libwget_la-ssl_openssl.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
ar: `u' modifier ignored since `D' is the default (see `U')
make[2]: Leaving directory '/mnt/mss/stuff/techy-bits/work/table/wget2-1.99.2/libwget'
make[1]: *** [Makefile:1700: all-recursive] Error 1
make[1]: Leaving directory '/mnt/mss/stuff/techy-bits/work/table/wget2-1.99.2'
make: *** [Makefile:1608: all] Error 2
```

**Comment by rockdaboot** [comment]
Nobody yet tried to get libressl support into wget2.

`--with-ssl=libressl`: Unknown strings are obviously just ignored without error message. That should be fixed.

`--with-openssl` just selects the crypto library (libcrypto, part of openssl). This has nothing to do with TLS, so GnuTLS is taken (as default).

The errors you see are likely due to an old openssl version, or - if you have libressl installed instead - the openssl compatibility layer is incomplete :man_shrugging: Maybe you can give some more information about type + version of your installed TLS library.

**Comment by rockdaboot** [comment]
Just made 2 commits to master for both issues. Tested with LibreSSL 3.0.2. You have to configure with `--with-ssl=openssl`.

Thanks for your report !

**Comment by ernstki** [comment]
On a CentOS 7 system, `./configure --with-ssl=openssl` and LibreSSL 3.4.2 followed by `make` fails with warnings and errors along the lines of

```
error : implicit declaration of strlcpy
```

For those just learning about [`strlcpy`](https://linux.die.net/man/3/strlcpy) today—like me—it's a string manipulation function from BSD, and apparently [a very contentious one](https://lwn.net/Articles/612244/). It's not part of the GNU C standard library, as far as I'm aware, at least not the system-default GCC of CentOS 7 vintage.

The problem here is that `strlcpy` _is_ provided by LibreSSL's libcrypto, so `./configure` finds it and sets `HAVE_STRLCPY=1`, which plays a role in [`libwget/strlcpy.c`](https://github.com/rockdaboot/wget2/blob/13aa6dea3ffc54935325d549fc7b372d1c5d958a/libwget/strlcpy.c#L57).

However, `libwget/strlcpy.c` doesn't include the necessary header file to pull in the [function prototype](https://github.com/libressl-portable/portable/blob/a27642bf418cb1d464d2d19226e5a13007068cf6/include/compat/string.h#L34). It would need to do something like `#include <compat/string.h>`, with the include path having been set to something like `/path/to/libressl/include`. Hello `CFLAGS=-I …`.

I didn't have the patience for all that, so I built OpenSSL 1.1.1m, then built wget 2.0.0 against that, which got me a _little_ farther. Just including this here for posterity, hopefully to save somebody some trouble.
